go-controller/pkg/clustermanager/userdefinednetwork/notifier/net-attach-def.go

```diff
@@ -0,0 +1,61 @@
+package notifier
+
+import (
+	"sync"
+
+	"k8s.io/client-go/util/workqueue"
+
+	netv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	netv1infomer "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/informers/externalversions/k8s.cni.cncf.io/v1"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/controller"
+)
+
+// NetAttachDefNotifier watches NetworkAttachmentDefinition objects and notify subscribers upon change.
+// It enqueues the reconciled object keys in the subscribing controllers workqueue.
+type NetAttachDefNotifier struct {
+	lock sync.Mutex
+
+	Controller controller.Controller
+
+	subscribers []controller.Controller
+}
+
+func NewNetAttachDefNotifier(nadInfomer netv1infomer.NetworkAttachmentDefinitionInformer, subscribers ...controller.Controller) *NetAttachDefNotifier {
+	c := &NetAttachDefNotifier{
+		subscribers: subscribers,
+	}
+
+	nadLister := nadInfomer.Lister()
+	cfg := &controller.ControllerConfig[netv1.NetworkAttachmentDefinition]{
+		RateLimiter:    workqueue.DefaultControllerRateLimiter(),
+		Reconcile:      c.Reconciler,
+		ObjNeedsUpdate: c.needUpdate,
+		Threadiness:    1,
+		Informer:       nadInfomer.Informer(),
+		Lister:         nadLister.List,
+	}
+	const nadControllerName = "user-defined-network-net-attach-def-controller"
+	c.Controller = controller.NewController[netv1.NetworkAttachmentDefinition](nadControllerName, cfg)
+
+	return c
+}
+
+func (c *NetAttachDefNotifier) needUpdate(_, _ *netv1.NetworkAttachmentDefinition) bool {
+	return true
+}
+
+func (c *NetAttachDefNotifier) Reconciler(key string) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	for _, subscriber := range c.subscribers {
+		if subscriber != nil {
+			// enqueue the reconciled NAD key in the subscribers workqueue to
+			// enable the subscriber act on NAD changes (e.g.: reflect NAD state is status)
+			subscriber.Reconcile(key)
```

Comment:
> I don't think Reconcile is the right method to call, as in the UDN controller example, that will interpret NAD key as UDN key, but NAD doesn't have namespace in the key (ever if it had, still doesn't look right). 

NAD is namesapced CRD, why wont the NAD object have namespace part of its key? 

So you can probably call this method ReconcileNAD, and make sure UDN Controller handles it correctly.

The controller Reconcile method is called here in order to enqueue NAD object changes to the UDN controller workqueue so it could act on NAD changes.
The UDN controller reconciler is implemented in a way it handle the key properly.
It fetch the UDN and NAD object from the cache using the key, which is the same for UDN and NAD objects.
Then reconcile the state accodingly (e.g.: if NAD is not found it should create it).